### PR TITLE
Remove NetworkId

### PR DESCRIPTION
**Pull request by jiexi**
## Explanation
Wallets shouldn't be directly concerned about the network ID as this more of a p2p concept for gossip. What wallets really care about is chain ID as that is the correct value to use to identify a chain, build transactions, etc. Although these two values usually match (ignoring hex/dec formatting), there are [exceptions](https://medium.com/@pedrouid/chainid-vs-networkid-how-do-they-differ-on-ethereum-eec2ed41635b).

We want to remove usage of networkId and replace it with chainId where appropriate and necessary. 

This was a generally straight forward change to make, except for the following cases:
* `networkVersion` on the `window.ethereum` provider object
* `metamaskNetworkId` on TransactionMeta objects

~~In both cases, we now use static mappings to assist in covering chainId <-> networkId edge cases.~~ See comments in this PR for more elaboration and the core PR linked below.

* Fixes [mmp-1068](https://github.com/MetaMask/MetaMask-planning/issues/1068)
* See: [core-1633](https://github.com/MetaMask/core/pull/1633)
* See: https://github.com/MetaMask/smart-transactions-controller

## Pre-merge author checklist

- [x] I've clearly explained:
  - [x] What problem this PR is solving
  - [x] How this problem was solved
  - [x] How reviewers can test my changes
- [x] Sufficient automated test coverage has been added

## Pre-merge reviewer checklist

- [x] Manual testing (e.g. pull and build branch, run in browser, test code being changed)
- [x] PR is linked to the appropriate GitHub issue
- [ ] **IF** this PR fixes a bug in the release milestone, add this PR to the release milestone

If further QA is required (e.g. new feature, complex testing steps, large refactor), add the `Extension QA Board` label.

In this case, a QA Engineer approval will be be required.

### CODE REVIEW

**Comment by jiexi** [comment]
@metamaskbot update-policies

**Comment by jiexi** [comment]
We expose `networkVersion` on the provider via `window.ethereum.networkVersion`. As always, since this has been exposed to the world, some dapps may rely on it. [EIP1193](https://eips.ethereum.org/EIPS/eip-1193) does not specify exposing this as a requirement of spec. We have a few options:
1. Remove it. This is a breaking change, but can be worked around by making an async `net_version` request
2. Keep it. We still need to make the request for `net_version` ourselves.
3. We can do that in `MetaMaskController` (cache it, making only one request per chain change event) OR
4. We can keep `NetworkController.networkId` around, but rename it to `deprecatedNetworkId` to make it very obvious developers should not be using it internally, and of course we will keep the changes in this PR that actual replace/remove its usage in extension

**Comment by jiexi** [comment]
Phasing out `metamaskNetworkId` from `Transaction` requires us to make a decision on how to handle legacy transactions without `chainId` populated.

We cannot naively convert Network ID => Chain ID via `toHex()` because Network ID may not always match the value of Chain ID (see ETC, where networkID=1, but chainID=61).

Possible options:
1. Build a map of networkId => chainIds. Use that to compare current chain ID against a transaction's networkId in the case that the transaction does not have chainId populated
2. Build a map of networkId => chainIds. Migrate transaction state so that each transaction with only `metamaskNetworkId` populated gets turned into multiple transactions with `chainId` set to each possible value it the networkId could correspond to. Drop `metamaskNetworkId` afterwards
3. Only filter by Transaction.chainId. Filter all transactions that do not have `chainId` set
4. Continue to pass `networkId` into `TransactionController`, but via a constructor params func instead of being read off the state object.

**Comment by jiexi** [comment]
Action items after team discussion:

* Deprecation warning on networkVersion (and other relevant fields. Handle this in a separate PR)
* Make net_version call in MetaMaskController each time network changes
* Pass net_version result from MMC to Transaction Controller

**Comment by jiexi** [comment]
@metamaskbot update-policies

File: package.json
Comment (jiexi) [inline package.json:1]: TODO: needs to be updated for core and smart-transaction-controller once they are merged

File: app/scripts/metamask-controller.test.js
Comment (jiexi) [inline app/scripts/metamask-controller.test.js:1]: TODO: Update this after https://github.com/MetaMask/metamask-extension/pull/20696 is merged

File: ui/components/app/transaction-activity-log/transaction-activity-log.util.js [hunk 1]
```diff
@@ -54,7 +54,6 @@ export function getActivities(transaction, isFirstTransaction = false) {
   const {
     id,
     chainId,
-    metamaskNetworkId,
```
Comment (jiexi) [inline ui/components/app/transaction-activity-log/transaction-activity-log.util.js:57]: this may need to stay here. thoughts?

File: app/scripts/metamask-controller.test.js
Comment (jiexi) [reply app/scripts/metamask-controller.test.js:1]: done

**Comment by jiexi** [comment]
@metamaskbot update-policies

**Comment by jiexi** [comment]
Determined that for transactions, chain ID has always existed on either TransactionMeta.chainId or TransactionMeta.txParams.chainId. We check both locations now and removed the fallback to checking TransactionMeta.networkId

Since we don't need the mapping of networkId to chainIds anymore, I've also removed the reversed chainId to networkId mapping and now make `net_version` calls in MetaMaskController to populate `networkVersion` again.
Release to add deprecation warnings is [here](https://github.com/MetaMask/providers/pull/283)

**Comment by jiexi** [comment]
@metamaskbot update-policies

File: app/scripts/controllers/transactions/index.js [hunk 2]
```diff
@@ -874,26 +871,13 @@ export default class TransactionController extends EventEmitter {
...
-      // is loading for two reasons.
-      // 1. Pending, unconfirmed transactions are wiped on network change
-      // 2. The UI is unusable (loading indicator) when network is loading.
-      // setting the networkId to 0 is for type safety and to explicity lead
-      // the transaction to failing if a user is able to get to this branch
-      // on a custom network that requires valid network id. I have not ran
-      // into this limitation on any network I have attempted, even when
-      // hardcoding networkId to 'loading'.
-      networkId:
-        networkStatus === NetworkStatus.Available ? parseInt(networkId, 10) : 0,
```
Comment (matthewwalsh0) [inline app/scripts/controllers/transactions/index.js:896]: Are we not losing functionality here if we're no longer concerned with the network status?

Should we use the `_getChainId` method instead as that already factors the `NetworkStatus`, converts to a decimal, and even returns zero as this removed logic did.

File: app/scripts/controllers/transactions/tx-state-manager.js [hunk 3]
```diff
@@ -278,9 +272,9 @@ export default class TransactionStateManager extends EventEmitter {
     const txsToDelete = transactions
       .reverse()
       .filter((tx) => {
-        const { nonce, from } = tx.txParams;
-        const { chainId, metamaskNetworkId, status } = tx;
-        const key = `${nonce}-${chainId ?? metamaskNetworkId}-${from}`;
+        const { chainId, status, txParams } = tx;
+        const txChainId = chainId ?? txParams.chainId;
```
Comment (matthewwalsh0) [inline app/scripts/controllers/transactions/tx-state-manager.js:276]: This fallback isn't required as both are set using the value from the network state, plus the `chainId` is stripped from the `txParams` during normalisation.

File: app/scripts/controllers/transactions/index.js (hunk 2 above)
Comment (jiexi) [reply app/scripts/controllers/transactions/index.js:896]: afaik, networkId isn't used in any part of the unsignedTx generation.

I see that in the next version of @ethereumjs/common (which we currently are behind) the common.Custom class's networkId param is [explicitly checked](https://github.com/ethereumjs/ethereumjs-monorepo/blob/acd1bbbcb436f8e3f137e06b7f85271424e5cd09/packages/common/src/common.ts#L271C8-L271C8) even though I don't see it being used to [build the unsigned transaction](https://github.com/ethereumjs/ethereumjs-monorepo/blob/master/packages/tx/src/transactionFactory.ts#L29C17-L29C27), which is what I expect. I don't understand why networkId on the Custom class is being enforced now. I guess we can just pass any value for networkId here. Maybe I'm missing something? If network ID is truly required by @ethereumjs libs for some reason, this could also force us to keep it around.

Comment (jiexi) [reply app/scripts/controllers/transactions/index.js:896]: We should be okay here

**Comment by jiexi** [comment]
@metamaskbot update-policies

File: app/scripts/metamask-controller.js [hunk 4]
```diff
@@ -385,7 +387,7 @@ export default class MetamaskController extends EventEmitter {
             rpcPrefs: {},
             rpcUrl: 'http://localhost:8545',
             ticker: 'ETH',
-            networkConfigurationId: 'networkConfigurationId',
+            id: 'networkConfigurationId',
```
Comment (jiexi) [inline app/scripts/metamask-controller.js:390]: These initialNetworkControllerState changes are to accommodate unreleased NetworkController changes that are being brought into the preview build that is used in this branch. They are unrelated to removing networkId, will likely be addressed in a separate PR to extension, and can be ignored for now

Comment (jiexi) [reply app/scripts/metamask-controller.js:397]: These NetworkController changes have just been released in https://github.com/MetaMask/core/pull/1697 and should be adopted in extension via separate PR.

**Comment by jiexi** [comment]
Still need to get https://github.com/MetaMask/smart-transactions-controller/pull/202 in

**Comment by jiexi** [comment]
@metamaskbot update-policies

File: app/scripts/controllers/transactions/tx-state-manager.js [hunk 5]
```diff
@@ -279,9 +272,8 @@ export default class TransactionStateManager extends EventEmitter {
     const txsToDelete = transactions
       .reverse()
       .filter((tx) => {
-        const { nonce, from } = tx.txParams;
-        const { chainId, metamaskNetworkId, status } = tx;
-        const key = `${nonce}-${chainId ?? metamaskNetworkId}-${from}`;
+        const { chainId, status, txParams } = tx;
+        const key = `${txParams.nonce}-${chainId}-${txParams.from}`;
```
Comment (mcmire) [inline app/scripts/controllers/transactions/tx-state-manager.js:276]: It seems like this would be one of the most impactful lines in this PR. What are the possible effects of this change?

File: ui/ducks/domains.js [hunk 6]
```diff
@@ -143,17 +142,17 @@ export function initializeDomainSlice() {
   return (dispatch, getState) => {
     const state = getState();
     const chainId = getCurrentChainId(state);
-    const network = CHAIN_ID_TO_NETWORK_ID_MAP[chainId];
-    const networkName = NETWORK_ID_TO_ETHERS_NETWORK_NAME_MAP[network];
-    const ensAddress = networkMap[network];
+    const networkName = CHAIN_ID_TO_ETHERS_NETWORK_NAME_MAP[chainId];
+    const chainIdInt = parseInt(chainId, 16);
```
Comment (mcmire) [inline ui/ducks/domains.js:146]: This could return `NaN`, is it worth checking?

File: ui/hooks/useTransactionInsights.js [hunk 7]
```diff
@@ -40,8 +38,7 @@ const useTransactionInsights = ({ txData }) => {
   );
 
   const { txParams, chainId, origin } = txData;
-  const networkId = CHAIN_ID_TO_NETWORK_ID_MAP[chainId];
-  const caip2ChainId = `eip155:${networkId ?? stripHexPrefix(chainId)}`;
+  const caip2ChainId = `eip155:${parseInt(chainId, 16).toString()}`;
```
Comment (mcmire) [inline ui/hooks/useTransactionInsights.js:41]: Do we need to be worried about `parseInt` returning NaN?

**Review by mcmire — COMMENTED** [review]
This is a heavy PR 😅 but I recognize many of these files from when we were making changes to NetworkController. We were going to remove network ID but decided not to do that then as we surmised, correctly I see, that it would be a lot of work to fully remove the network ID. It looks like we would have missed some things if we'd attacked it then, so good job on doing this thoroughly. I guess this assumes that all chains support chain IDs? I mean they _ought_ to, but is there any reason why we shouldn't do this?

File: app/scripts/controllers/transactions/tx-state-manager.js (hunk 5 above)
Comment (jiexi) [reply app/scripts/controllers/transactions/tx-state-manager.js:276]: if there are greater than 60 of transactions that do not have `txMeta.chainId` populated, only the most recent 60 of those transactions will be kept. The older ones are purged from state. This would mirror the new behavior in [the core TransactionController ](https://github.com/MetaMask/core/blob/5e1044cd02180ad9bfe574a9ec4096b13ca0efbd/packages/transaction-controller/src/TransactionController.ts#L1379C17-L1381).

File: ui/ducks/domains.js (hunk 6 above)
Comment (jiexi) [reply ui/ducks/domains.js:146]: honestly not sure... I don't check if `parseInt` is returns `NaN` because it's the pattern I've generally seen used elsewhere in extension. For example, see line 151 in this file

**Comment by jiexi** [comment]
@metamaskbot update-policies

File: app/scripts/controllers/transactions/tx-state-manager.js (hunk 5 above)
Comment (mcmire) [reply app/scripts/controllers/transactions/tx-state-manager.js:276]: Makes sense!

**Comment by pedronfigueiredo** [comment]
Should we include a migration to remove `networkId` from the `NetworkController` state?

**Comment by adonesky1** [comment]
@metamaskbot update-policies

**Comment by jiexi** [comment]
@metamaskbot update-policies

**Comment by jiexi** [comment]
> Should we include a migration to remove networkId from the NetworkController state?

Yeah, we should be tidy. My bad. Added here: https://github.com/MetaMask/metamask-extension/pull/20652/commits/1c3bb6605b377506d93a8411b9df4c91f25781a1

**Review by pedronfigueiredo — APPROVED** [review]

**Comment by jiexi** [comment]
> I guess this assumes that all chains support chain IDs? I mean they ought to, but is there any reason why we shouldn't do this?

All EVM chains support chain ID. I'm unsure of other chains. I can't think of a good reason not to do this besides not changing things if they aren't broken. This doesn't buy us any new features, just cleans up some tech debt

File: app/scripts/controllers/ens/ens.js [hunk 8]
```diff
@@ -1,17 +1,18 @@
...
+  static getChainEnsSupport(chainId) {
+    return Boolean(ensNetworkMap[parseInt(chainId, 16).toString()]);
   }
 
-  constructor({ network, provider } = {}) {
-    const networkName = NETWORK_ID_TO_ETHERS_NETWORK_NAME_MAP[network];
-    const ensAddress = ensNetworkMap[network];
+  constructor({ chainId, provider } = {}) {
+    const networkName = CHAIN_ID_TO_ETHERS_NETWORK_NAME_MAP[chainId];
+    const chainIdInt = parseInt(chainId, 16);
```
[the remark this answers is not in the document]
Comment (adonesky1) [reply app/scripts/controllers/ens/ens.js:12]: seems like this question has been addressed in a separate thread?

File: ui/hooks/useTransactionInsights.js (hunk 7 above)
Comment (adonesky1) [reply ui/hooks/useTransactionInsights.js:41]: seems like this question has been [addressed in a separate thread?](https://github.com/MetaMask/metamask-extension/pull/20652#discussion_r1349304037)

File: package.json [hunk 9]
```diff
@@ -275,7 +275,7 @@
     "@metamask/rpc-methods": "^3.0.0",
     "@metamask/safe-event-emitter": "^2.0.0",
     "@metamask/scure-bip39": "^2.0.3",
-    "@metamask/selected-network-controller": "^1.0.0",
+    "@metamask/selected-network-controller": "^2.0.0",
```
Comment (adonesky1) [inline package.json:278]: [nit] Just curious, is this required for this PR?

**Review by adonesky1 — APPROVED** [review]
LGTM. Obviously a ton of the same little changes here so hard to feel super confident without spending many hours pouring over each one in context. So hoping we can get at least a few more sets of eyes and some thorough QA done before we merge this.

**Comment by adonesky1** [comment]
@tmashuang do you have some bandwidth to give this a thorough QA?

File: package.json (hunk 9 above)
Comment (jiexi) [reply package.json:278]: selected-network-controller v1.0.0 had a peerDep of network-controller ^12.1.2.

v2.0.0 has a peerDep of 13

**Comment by jiexi** [comment]
Thomas and I took another pass on this. I think we're good on the QA side
